Migration docs (DCNE-321) (#1322)
CiscoDevNet/terraform-provider-aci: docs/guides/migration.md

---
subcategory: "Guides"
page_title: "Migration"
description: |-

>     An overview of resource and datasource migration changes

Reviewer: Can we add migration from Terraform SDKv2 to Plugin Framework?

Reviewer: or just migration to the latest Terraform Plugin Framework?

Author: removed latest but went with your second suggestion and also added to title

---

## Introduction

> Since the first release in September 2020, the Terraform ACI provider has come a long way, adding new resources and datasources to make managing Cisco ACI environments easier. Over the years, we've moved from SDKv1 to [SDKv2](https://developer.hashicorp.com/terraform/plugin/sdkv2), and now to the latest [Terraform Plugin Framework](https://developer.hashicorp.com/terraform/plugin/framework), which is the recommended way to develop Terraform plugins and offers many advantages over SDKv2.

Reviewer: Since the first release -> Since its first release

Reviewer: datasources -> data sources ?

Author: changed everywhere

To take full advantage of the new features in the Terraform Plugin Framework, we had to completely rewrite the Terraform ACI provider resources and datasources. While this was a big task, it also gave us a chance to review and improve the current ACI provider. This guide explains the new features that the Terraform Plugin Framework provides and the changes we have made to the ACI provider and why we made them.

## Upgrading the ACI Terraform Provider

Upgrading the ACI Terraform provider to a new version involves several steps to ensure a smooth transition and to avoid any disruptions to your existing configurations. This guide will walk you through the process, including taking a backup of your current state.

#### Step 1: Backup Your Current State

Before making any changes, it's crucial to back up your current Terraform state. This ensures that you can restore your environment if anything goes wrong during the upgrade process.

> 1. **Local Backend**: Navigate to Your Terraform Project Directory and open a terminal and navigate to the directory containing your Terraform configuration files and copy the current state file (`terraform.tfstate`) to a backup location.

Reviewer: Navigate to Your Terraform Project Directory and open a terminal and navigate to the directory containing your Terraform configuration files and copy the current state file (`terraform.tfstate`) to a backup location.
->
Open a terminal and navigate to the directory containing your ACI Terraform configuration files with the state and copy the current state file (`terraform.tfstate`) to a backup location.

        "filename": "main.tf",
        "start_line": 72
    }
    ```

2. **Change the deprecated attributes**: Replace the deprecated attributes in your configuration file with the new attributes and execute the plan again.

    ***Old Configuration***
    ```hcl
    resource "aci_bridge_domain" "terraform_bd" {
        tenant_dn = aci_tenant.terraform_tenant.id
        name = "terraform_bd"
        ll_addr = "::"
        unk_mcast_act = "flood"
    }
    ```

    ***New Configuration***
    ```hcl
    resource "aci_bridge_domain" "terraform_bd" {
        parent_dn = aci_tenant.terraform_tenant.id
        name = "terraform_bd"
        link_local_ipv6_address = "::"
        l3_unknown_multicast_flooding = "flood"
    }
    ```

    The `terraform plan` command should not display any warnings anymore.

    ```bash
    No changes. Your infrastructure matches the configuration.

    Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
    ```

### Step 6: Cleanup

> 1. **Remove Backup (Optional)**: If everything is working correctly, you can remove the backup state file.

Reviewer: Does the colour of the font change? There are a few inconsistencies

Author: What do you mean by this? The colours displayed here in GitHub? 

I used the https://registry.terraform.io/tools/doc-preview tool and VsCode, here all the titles and bullet points seems consistent

## Changes to the ACI Provider

In this section, we outline the key changes made to the Terraform ACI provider as part of the migration to the Terraform Plugin Framework. These changes aim to enhance the provider's functionality, performance, and usability. The new provider is generated from meta files, ensuring consistency and accuracy across resources and datasources. Below, we detail the specific improvements and modifications implemented during this migration.

## Cleaning and Renaming of Attributes

Over time, some attributes became outdated, were inconsistently named, or were not exposed correctly. This made the resources sometimes challenging to understand and thus use the ACI provider effectively. By cleaning and renaming these attributes, we aim to:

1. **Improve Clarity**: Ensure that attribute names are clear and descriptive, making it easier for users to understand their purpose.
2. **Enhance Consistency**: Standardize attribute names and structures across resources and datasources.
3. **Simplify Usage**: Remove outdated attributes to streamline the configuration process.

> We understand the importance of maintaining backward compatibility to avoid disrupting existing configurations. This is why during a transition period, both old and new style attributes can be used. It allows users to continue using their existing configurations while gradually adopting the new style attributes. We have implemented deprecation warnings for old style attributes, informing users of the upcoming changes and encouraging them to transition to the new attribute style as soon as possible.

Reviewer: new style attributes -> appropriately redefined attributes?

Author: closing this one and commenting on your comment below

> > For the same ACI property, the old and new style attributes cannot be used together. Doing so will cause the provider to error during configuration validation.

Reviewer: Would be good to change this "new style attributes"

Author: Yes think we as the team should decide on how to name this, have no preference will change accordingly everywhere when decision is made. 

@sajagana @samiib @anvitha-jain @gmicol @lhercot , any suggestions?

Reviewer: Example:
![Screenshot 2025-03-25 at 10 27 17 AM](https://github.com/user-attachments/assets/247a0fb9-6222-4f28-ae12-11571d0c93bf)

Reviewer 2: Can we use "legacy and modern/new attributes" to replace the "old and new style attributes"?

Author: putting suggestion of @shrsr here since this was in separate comment new style attributes -> appropriately redefined attributes?

Author: changed

Finally, because the final plan needs to match the applied state, old style attributes that represent a DN of a named relation require to be resolved into the correct tDn. This means that the object must exist when creating the relationship; if you fail to do so, the provider will panic. This can be prevented by leveraging new style resources where the configurable attribute is used.

## Changed Child Objects in Configuration

In migrated resources, all children MOs are exposed in configuration as a map or a list of maps, instead of single attribute types (e.g., string, boolean, integer). The decision is made to have the configuration resemble the model more to provide flexibility of adding new attributes for a child without introducing breaking changes.

## Changed Child Objects in Payloads

In non-migrated resources, the children were managed via individual REST API requests, which could lead to a large number of REST API calls for a single resource when a lot of children were defined. For migrated resources, the decision is made to include all children inside a single GET/POST request to limit the number of REST API calls.

## Changed Annotation Behavior

> The annotation attribute can be set for each MO that exposes the attribute, including children inside a resource. By default, the annotation is set to `orchestrator:terraform`, but this default can be overwritten at the provider level.

Author: added an example below

To take full advantage of the new features in the Terraform Plugin Framework, we had to completely rewrite the Terraform ACI provider resources and data sources. While this was a big task, it also gave us a chance to review and improve the current ACI provider. This guide explains the new features that the Terraform Plugin Framework provides and the changes we have made to the ACI provider and why we made them.

## Upgrading the ACI Terraform Provider

Upgrading the ACI Terraform provider to a new version involves several steps to ensure a smooth transition and to avoid any disruptions to your existing configurations. This guide will walk you through the process, including taking a backup of your current state.

> #### Step 1: Backup Your Current State

Reviewer 2: ```suggestion
### Step 1: Backup Your Current State
```

## Changed Child Objects in Payloads

In non-migrated resources, the children were managed via individual REST API requests, which could lead to a large number of REST API calls for a single resource when a lot of children were defined. For migrated resources, the decision is made to include all children inside a single GET/POST request to limit the number of REST API calls.

## Changed Annotation Behavior

The annotation attribute can be set for each MO that exposes the attribute, including children inside a resource. By default, the annotation is set to `orchestrator:terraform`, but this default can be overwritten at the provider level.

```hcl
provider "aci" {
  username   = "admin"
  password   = "password"
  url        = "https://my-cisco-aci.com"
  annotation = "provider-level-annotation-overwrite"
}
```

## Showing ID in Plan for Create

For non-migrated resources, the ID of the resource appears as "known after apply." In the case of migrated resources, the ID is calculated during the planning phase and included in the plan output.

## Error for Existing MO on Create

> In migrated resources, we provide the provider level option (`allow_existing_on_create`) to check during the plan if an object already exists. By default, we allow existing MOs to be managed, but this option can be set to `false` to allow for a safeguard mechanism to prevent managing the same MO by different configuration parts. The drawback of this safeguard mechanism is that during a plan, an additional API call is made per resource to verify that the object does not already exist.

Reviewer 3: ```suggestion
Migrated resources offer a provider-level option (`allow_existing_on_create`) to check during the plan if an object already exists. By default, existing MOs can be managed, but this option can be set to `false`, providing a safeguard mechanism to prevent managing the same MO by different configuration parts. The drawback of this safeguard mechanism is an additional API call per resource during the plan to verify that the object does not already exist.
```

## Allowing Empty Input for Attributes

> Terraform supports three states for any value: null (missing), unknown ("known after apply"), and known. The previous SDKs did not support null and unknown value states, which meant that before the Plugin Framework, we couldn't differentiate between an empty value and a value that wasn't provided. Because of this limitation, we were unable to update a string value to an empty string.

Reviewer 3: ```suggestion
Terraform supports three states for any value: null (missing), unknown ("known after apply"), and known. Previous SDKs did not support null and unknown states, which prevented differentiating between an empty value and a non-provided value. Because of this limitation, updating a string value to an empty string was not possible before the Plugin Framework.
```

1. **Improve Clarity:** Ensure that attribute names are clear and descriptive, making it easier for users to understand their purpose.
2. **Enhance Consistency:** Standardize attribute names and structures across resources and data sources.
3. **Simplify Usage:** Remove outdated attributes to streamline the configuration process.

We understand the importance of maintaining backward compatibility to avoid disrupting existing configurations. Therefore, during the transition period, both legacy and redefined attributes are supported. This approach allows users to continue using their existing configurations while gradually adopting the redefined attributes. We have implemented deprecation warnings for legacy attributes, informing users of the upcoming changes and encouraging them to transition to the redefined attribute as soon as possible.

> It is important to note that for the same ACI property, legacy and redefined attributes cannot be used simultaneously. Attempting to do so will result in an error during configuration validation.

> A downside to this approach is increased verbosity in the plan output due to known after applies for each legacy attribute not provided when a change is detected. This is a temporary drawback which will be resolved once deprecated attributes are removed in the next major release.

Author: changed a bit different from your suggestion, please press resolve if this is ok for you

1. **Local Backend:** Open a terminal and navigate to the directory where your ACI Terraform configuration files with the state are located. Copy the current state file (terraform.tfstate) to a backup location..
   
   ```bash
   cd /path/to/your/terraform/project
   cp terraform.tfstate terraform.tfstate.backup
   ```

2. **Remote Backend:** If you are using a remote backend (e.g., S3, Azure Blob Storage), ensure you have a backup of the state file from the remote location.

### Step 2: Update the Provider Version

1. **Open the Terraform Configuration File:** Open the `main.tf` or the relevant Terraform configuration file where the ACI provider is defined.

2. **Update the Provider Version:** Modify the provider block to specify the new version of the ACI provider.

   ```hcl
   provider "aci" {
     version = "x.y.z"  # Replace with the new version number
     # Other provider configuration options
   }
   ```

3. **Initialize the Configuration:** Run `terraform init` to reinitialize your configuration with the updated provider version.

   ```bash
   terraform init

>    ```

Author: checked and changed the command to include -upgrade

        "filename": "main.tf",
        "start_line": 72
    }
    ```

2. **Change the deprecated attributes:** Replace the deprecated attributes in your configuration file with the redefined attributes and execute the plan again.

    ***Old Configuration***
    ```hcl
    resource "aci_bridge_domain" "terraform_bd" {
        tenant_dn     = aci_tenant.terraform_tenant.id
        name          = "terraform_bd"
        ll_addr       = "::"
        unk_mcast_act = "flood"
    }
    ```

    ***New Configuration***
    ```hcl
    resource "aci_bridge_domain" "terraform_bd" {
        parent_dn                     = aci_tenant.terraform_tenant.id
        name                          = "terraform_bd"
        link_local_ipv6_address       = "::"
        l3_unknown_multicast_flooding = "flood"
    }
    ```

    The `terraform plan` command should not display any warnings anymore.

    ```bash
    No changes. Your infrastructure matches the configuration.

    Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
    ```

> 

Author: adjusted the sentence above

## Optional Optimization Possibilites

1. **Add relationships to parent resource:** Reduces the amount of resources in the configuration by including children inside the parent config. This will decrease the execution time by decreasing the size of the terraform graph and reducing the amount of REST API calls made towards APIC.

    ***Old Configuration***
    ```hcl
    resource "aci_application_epg" "terraform_epg" {
        parent_dn = aci_application_profile.terraform_ap.id
        name      = "terraform_epg"
    }

    resource "aci_epg_to_domain" "terraform_epg_to_domain" {
        application_epg_dn = aci_application_epg.terraform_epg.id
        tdn                = "uni/phys-phys"
    }
    ```

    ***New Configuration***
    ```
    resource "aci_application_epg" "terraform_epg" {
        parent_dn = aci_application_profile.terraform_ap.id
        name      = "terraform_epg"
        relation_to_domains = [
            {
                target_dn = "uni/phys-phys"
            }
        ]
    }
    ```

>     !> Child resources should not be used in combination with nested attributes in parent resources. Doing so will result in unexpected behaviour.

Reviewer 4: I think we need to add a sentence here that is clearer like: Use one or the other, Either use a child or add it in the parent or something similar.

Author: ok will also add a note for the redesign to rephrase, because this sentence was previously defined and used already in resource docs

>     In this scenario the child configuration should not be destroyed, but should be unmanaged (removed from state) before executing the plan. Terraform will display destroy intent in the `terraform plan` output when executed without unmanaging the child resource first.

Reviewer 4: ```suggestion
    In this scenario the child configuration should not be destroyed, but should be unmanaged (removed from state) before executing the plan. Terraform will display destroy intent in the `terraform plan` output when executed without unmanaging the child resource first.
```
What is the scenario? Maybe describe the scenario before showing the example or rephrase this. Also, add a sentence or link to the next part when you show how to unmanage.

    # (because aci_epg_to_domain.terraform_epg_to_domain is not in configuration)
    - resource "aci_epg_to_domain" "terraform_epg_to_domain" {
      - application_epg_dn    = "uni/tn-terraform_tenant/ap-terraform_ap/epg-terraform_epg" -> null
      - tdn                   = "uni/phys-phys" -> null
      ...
      - vmm_id                = jsonencode({})
        # (7 unchanged attributes hidden)
    }

    Plan: 0 to add, 0 to change, 1 to destroy.
    ```

    The [terraform state rm](https://developer.hashicorp.com/terraform/cli/commands/state/rm) command `terraform state rm aci_epg_to_domain.terraform_epg_to_domain` should be used to unmanage without destroying a resource. The output should reflect the succesful removal of the resource from state.

    ```bash
    Removed aci_epg_to_domain.terraform_epg_to_domain
    Successfully removed 1 resource instance(s).
    ```

    The `terraform plan` output should reflect no changes, when the resource is unmanaged.
    
    ```bash
    No changes. Your infrastructure matches the configuration.

    Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
    ```

>     The state file does not reflect the changes yet, because a refresh has not taken place. This can be done with the `terraform refresh` or `terraform apply -refresh-only` commands, which do not modify the objects in apic, but only modifies the state file. For more information see [refresh](https://developer.hashicorp.com/terraform/cli/commands/refresh) and [planning-modes](https://developer.hashicorp.com/terraform/cli/commands/plan#planning-modes) documentation of Terraform.

Reviewer 4: ```suggestion
    The state file does not reflect the changes yet, because a refresh has not taken place. This can be done with the `terraform refresh` or `terraform apply -refresh-only` commands, which do not modify the objects in APIC, but only modifies the state file. For more information see [refresh](https://developer.hashicorp.com/terraform/cli/commands/refresh) and [planning-modes](https://developer.hashicorp.com/terraform/cli/commands/plan#planning-modes) documentation of Terraform.
```

1. **Improve Clarity:** Ensure that attribute names are clear and descriptive, making it easier for users to understand their purpose.
2. **Enhance Consistency:** Standardize attribute names and structures across resources and data sources.
3. **Simplify Usage:** Remove outdated attributes to streamline the configuration process.

We understand the importance of maintaining backward compatibility to avoid disrupting existing configurations. Therefore, during the transition period, both legacy and redefined attributes are supported. This approach allows users to continue using their existing configurations while gradually adopting the redefined attributes. We have implemented deprecation warnings for legacy attributes, informing users of the upcoming changes and encouraging them to transition to the redefined attribute as soon as possible.

> It is important to note that for the same ACI property, legacy and redefined attributes cannot be used simultaneously. Attempting to do so will result in an error during configuration validation.

> A downside to this approach is increased verbosity in the plan output due to "known after applies" for each legacy attribute not provided when a change is detected. This is a temporary drawback which will be resolved once deprecated attributes are removed in the next major release.

Reviewer 5: ```suggestion
A downside to this approach is increased verbosity in the plan output due to the "known after apply" state for each legacy attribute not provided when a change is detected. This is a temporary drawback which will be resolved once deprecated attributes are removed in the next major release.
```

> A downside to this approach is increased verbosity in the plan output due to the "known after apply" state for each legacy attribute not provided when a change is detected. This is a temporary drawback which will be resolved once deprecated attributes are removed in the next major release.

Reviewer 4: ```suggestion
A downside to this approach is the increased verbosity in the plan output when a change is detected due to the display of the "known after apply" state for each non-provided legacy attribute. This is a temporary drawback which will be resolved once deprecated attributes are removed in the next major release.
```

## Changed Behavior for Relations

In an ACI setup, Managed Objects (MOs) represent the different physical and logical parts within the Management Information Tree (MIT). These MOs can be linked through relationship MOs, which define how different MOs are connected. Relationship MOs act as connectors that establish links between two or more MOs, helping to organize and structure the hierarchical relationships within the MIT. In ACI, these relationships can be of two types: explicit or named.

> 1. **Explicit relations** These require the target Distinguished Name (DN) to be specified in the tDn attribute of the relationship MO, where only one target exists for the relationship. If the target DN is absent, the relationship cannot be established.

Reviewer 4: ```suggestion
1. **Explicit relationships** These relationships require the target Distinguished Name (DN) to be specified in the tDn attribute of the relationship MO, where only one target exists for the relationship. If the target DN is absent, the relationship cannot be established.
```

> 2. **Named relations** These require the name (identifier) attribute to be specified in the relationship MO, triggering a resolving mechanism based on precedence order. This allows the relationship to form with any MO in the precedence order.

Reviewer 4: ```suggestion
2. **Named relationships** These relationships require the name (identifier) attribute to be specified in the relationship MO, triggering a resolving mechanism based on a precedence order. This allows the relationship to form with the first MO that exists in the precedence order.
```

> In non-migrated resources of the Terraform Provider ACI, the relationship types are hidden from the user by allowing a DN (resource ID) or name input. The relationship type determines which attribute (name vs tDn) is added to the payload. In SDKv2, there was no enforcement (only warnings in logs) of the final plan needing to match the applied state, which allowed us to strip the name from a provided DN for named relations and add that to the payload for named relations. However, the migration to the Plugin Framework enforces the final plan to match the applied state.

Reviewer 4: ```suggestion
In non-migrated resources of the Terraform Provider ACI, the relationship types are hidden from the user by allowing a DN (resource ID) or name input. The relationship type determines which attribute (name vs tDn) is added to the payload. In SDKv2, there was no enforcement (only warnings in logs) of the final plan needing to match the applied state, which allowed us to extract the name from a provided DN for named relationships and use that extracted name in the relationship payload. However, the Plugin Framework enforces that the final plan values have to match the applied state.
```

> Consequently, for the ACI Provider, using the name of the provided tDn for named relation MOs means the DN input may not match the resolved tDn, potentially causing provider errors. Due to this change, we decided to expose only configurable attributes as input for resources, requiring the name attribute instead of the DN for named relationships.

Reviewer 4: ```suggestion
Consequently, for the ACI Provider, using the name of the provided tDn for the named relationship MOs means that the user provided DN may not match the resolved tDn, potentially causing provider errors. Due to this change, we decided to expose only configurable attributes as input for resources, requiring the name attribute instead of the DN for named relationships.
```

> To ensure the final plan matches the applied state, legacy attributes representing a Distinguished Name (DN) of a named relation must be resolved into the correct target Distinguished Name (tDn). This requires the object to exist when establishing the relationship; failure to do so will cause the provider to panic. See provider panic exmaple below.

Reviewer 4: ```suggestion
To ensure the final plan matches the applied state, legacy attributes representing a Distinguished Name (DN) of a named relationship must be resolved into the correct target Distinguished Name (tDn). This requires the object to exist when establishing the relationship; the absence of the object will cause the provider to panic. See provider panic example below.
```

│ unexpected new value: .relation_fv_rs_bd_to_ep_ret: was cty.StringVal("uni/tn-terraform_tenant/epRPol-ep_pol"), but now
│ cty.StringVal("uni/tn-common/epRPol-default").
╵
```

This issue can be avoided by using the redefined resources with a configurable name attribute or by ensuring that the tDn can be resolved during the apply phase.

## Changes to Child Objects in Configuration

In migrated resources, all child Managed Objects (MOs) are represented in the configuration as a map or a list of maps, rather than single attribute types (such as string, boolean, or integer). This approach ensures the configuration closely resembles the model, allowing for the addition of redefined attributes to a child without causing breaking changes.

## Changed Child Objects in Payloads

In non-migrated resources, child objects were managed via individual REST API requests, which could lead to a large number of REST API calls for a single resource when many children were defined. For migrated resources, the decision is made to include all children inside a single GET/POST request to limit the number of REST API calls.

## Changed Annotation Behavior

> The annotation attribute can be set for each MO that exposes the attribute, including children inside a resource. By default, the annotation is set to `orchestrator:terraform`, but this default can be overwritten at the provider level.

Author: added an example for where you can overwrite the annotation? not sure what we should add as additional explanation

  annotation = "provider-level-annotation-overwrite"
}
```

## Showing ID in Plan for Create

For non-migrated resources, the ID of the resource appears as "known after apply". In the case of migrated resources, the ID is calculated during the planning phase and included in the plan output.

## Error for Existing MO on Create

Migrated resources offer a provider-level option (`allow_existing_on_create`) to check during the plan if an object already exists. By default, existing MOs can be managed, but this option can be set to `false`, providing a safeguard mechanism to prevent managing the same MO by different configuration parts. The drawback of this safeguard mechanism is an additional API call per resource during the plan to verify that the object does not already exist.

## Allowing Empty Input for Attributes

Terraform supports three states for any value: null (missing), unknown ("known after apply"), and known. Previous SDKs did not support null and unknown states, which prevented differentiating between an empty value and a non-provided value. Because of this limitation, updating a string value to an empty string was not possible before the Plugin Framework.

> 

Reviewer 4: I would add a clear statement that it is now supported. It currently let the user assume it is supported but we should be clearer.

## Documentation Enhancements

The documentation for migrated resources and data sources is automatically generated to enhance relevancy and consistency. The following changes have been made:

1. Class names have references to the [model documentation](https://developer.cisco.com/site/apic-mim-ref-api/?version=latest).
2. Parents of the resource are provided with references to their corresponding resources at the `parent_dn` attribute.
3. Relational children of the resource include references to both their corresponding child resource and the target resource they are associated with.
4. Non-relational children of the resource are provided with references to their corresponding resources at the bottom of the documentation.
5. The default behaviour of each attribute is documented individually.
6. The valid inputs for each attribute are documented individually.
7. Applicable versions of classes and attributes are documented.
8. UI location information is provided for each configuration option.
9. Resource DN format (resource ID) is documented.
10. Both minimal and full examples are provided, featuring up to two parent resources when applicable.

> 11. Examples for importing resources using both CLI commands and HCL configuration blocks are provided.

Author: will add

## Upgrading the ACI Terraform Provider

Upgrading the ACI Terraform provider to a new version requires careful planning to ensure a seamless transition and prevent any disruptions to your existing configurations. This guide will lead you through each step of the process, including how to back up your current state.

### Step 1: Backup Your Current State

Before making any changes, it's crucial to back up your current Terraform state. This precaution ensures that you can revert your environment to its previous state if any issues arise during the upgrade process.

1. **Local Backend:** Open a terminal and navigate to the directory where your ACI Terraform configuration files with the state are located. Copy the current state file (terraform.tfstate) to a backup location.
   
   ```bash
   cd /path/to/your/terraform/project
   cp terraform.tfstate terraform.tfstate.backup
   ```

2. **Remote Backend:** If you are using a remote backend (e.g., S3, Azure Blob Storage), ensure you have a backup of the state file from the remote location.

### Step 2: Update the Provider Version

> 

Author: added

1. **Open the Terraform Configuration File:** Open the `main.tf` or the relevant Terraform configuration file where the ACI provider is defined.

2. **Update the Provider Version:** We recommend setting [version constraints](https://developer.hashicorp.com/terraform/language/expressions/version-constraints) for the provider. Modify the provider block to specify the new version constraint for the ACI provider. 

   ***Version constraint format***
   ```hcl
   terraform {
    required_providers {
       aci = {
         source = "ciscodevnet/aci"
         version = "x.y.z"  # Replace with the new version number
       }
     }
   }
   ```

   ***Deprecated Terraform version constraint format***
   ```hcl
   provider "aci" {

>      version = "x.y.z"  # Replace with the new version number

Reviewer 4: ```suggestion
     version = "~> x.y.z"  # Replace with the new version number
```

---
subcategory: "Guides"
page_title: "Migration to Terraform Plugin Framework"
description: |-
    An overview of resource and data source migration to Terraform Plugin Framework
---

## Introduction

> Since its first release in September 2020, the Terraform ACI provider has significantly progressed, adding new resources and data sources to streamline the management of Cisco ACI environments. Over the years, we've transitioned from SDKv1 to [SDKv2](https://developer.hashicorp.com/terraform/plugin/sdkv2), and now to the latest [Terraform Plugin Framework](https://developer.hashicorp.com/terraform/plugin/framework). This new Plugin Framework is the recommended way for developing Terraform plugins and offers several advantages over SDKv2.

Author: ok

> In order to fully leverage the new features of the Terraform Plugin Framework, we undertook a complete rewrite of the Terraform ACI provider's resources and data sources. Although this was a significant effort, it presented an opportunity to review and enhance the existing ACI provider. This guide provides detailed steps for migrating from previous versions, outlines key optimization steps to enhance performance, and aims to facilitate a smoother transition for users adapting to these updates. Additionally, it outlines the new features introduced by the Terraform Plugin Framework, along with the modifications we implemented in the ACI provider and the rationale behind these changes.

Reviewer 4: ```suggestion
In order to fully leverage the new features of the Terraform Plugin Framework, we decided to undertake a complete rewrite of the Terraform ACI provider's resources and data sources. Although this is a significant effort, it presented an opportunity to review and enhance the existing ACI provider and how the provider is developed. 

This complete rewrite will stretch a series of releases starting with [v2.16.0](https://github.com/CiscoDevNet/terraform-provider-aci/releases/tag/v2.16.0) where all new resources and data sources will be implemented using the new Terraform Plugin Framework and pre-existing resources and data sources will be slowly migrated into the Terraform Plugin Framework. Migrated resources and data sources contain a note that specifies it has been migrated at the top of their documentation. Once all pre-existing resources are migrated to the Terraform Provider Framework, a major release of the provider will be created to remove legacy attributes.

This guide provides detailed steps for migrating from previous versions, outlines key optimization steps to enhance performance, and aims to facilitate a smoother transition for users adapting to these updates. Additionally, it outlines how the migrated resources use the new features introduced by the Terraform Plugin Framework, along with the modifications we implemented in the ACI provider and the rationale behind these changes.
```

## Upgrading the ACI Terraform Provider

Upgrading the ACI Terraform provider to a new version requires careful planning to ensure a seamless transition and prevent any disruptions to your existing configurations. This guide will lead you through each step of the process, including how to back up your current state.

### Step 1: Backup Your Current State

Before making any changes, it's crucial to back up your current Terraform state. This precaution ensures that you can revert your environment to its previous state if any issues arise during the upgrade process.

1. **Local Backend:** Open a terminal and navigate to the directory where your ACI Terraform configuration files with the state are located. Copy the current state file (terraform.tfstate) to a backup location.

   ```bash
   cd /path/to/your/terraform/project
   cp terraform.tfstate /path/to/your/backup/terraform.tfstate.backup

>    ```

Author: ok fix all indentation

    ```bash
    Removed aci_epg_to_domain.terraform_epg_to_domain
    Successfully removed 1 resource instance(s).
    ```

    The `terraform plan` output should reflect no changes, when the resource is unmanaged.
    
    ```bash
    No changes. Your infrastructure matches the configuration.

    Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
    ```

    The state file does not reflect the changes yet, because a refresh has not taken place. This can be done with the `terraform refresh` or `terraform apply -refresh-only` commands, which do not modify the objects in APIC, but only modifies the state file. For more information see [refresh](https://developer.hashicorp.com/terraform/cli/commands/refresh) and [planning-modes](https://developer.hashicorp.com/terraform/cli/commands/plan#planning-modes) documentation of Terraform.

> ## Changes to the ACI Provider

Author: changed let me know if this better

> ## Changes to the ACI Provider

Author: undid change after discussion

## Changed Annotation Behavior

The annotation attribute can be set for each MO that exposes the attribute, including children inside a resource. By default, the annotation is set to `orchestrator:terraform`, but this default can be overwritten at different levels (provider, resource, child attribute).

```hcl
provider "aci" {
  username   = "admin"
  password   = "password"
  url        = "https://my-cisco-aci.com"
  annotation = "provider-level-annotation-overwrite"
}

resource "aci_application_epg" "example" {
  parent_dn  = aci_application_profile.example.id
  name       = "test_name"
  annotation = "resource-level-annotation-overwrite"
  relation_to_bridge_domain = {
    annotation         = "child-level-annotation-overwrite"
    bridge_domain_name = aci_bridge_domain.example.name
  }
}
```

## Showing ID in Plan for Create

For non-migrated resources, the ID of the resource appears as "known after apply". In the case of migrated resources, the ID is computed during the planning phase and included in the plan output to reduce the cascading attributes displaying a "known after apply".

## Error for Existing MO on Create

> Migrated resources support the (`allow_existing_on_create`) provider-level option to check during the plan if an object already exists. By default, existing MOs will be automatically managed on apply, as (`allow_existing_on_create`) is set to `true` by default.  This option can be set to `false`, providing a mechanism to prevent the management of the same MO by separate configurations. The drawback of this mechanism is that it requires an additional API call per resource during the plan to verify that the object does not already exist.

Reviewer 4: ```suggestion
Migrated resources support the [`allow_existing_on_create`](https://registry.terraform.io/providers/CiscoDevNet/aci/latest/docs#allow_existing_on_create-1) provider-level option to check during the plan if an object already exists. By default, existing MOs will be automatically managed on apply, as `allow_existing_on_create` is set to `true` by default.  This option can be set to `false`, providing a mechanism to prevent the management of the same MO by separate configurations. The drawback of this mechanism is that it requires an additional API call per resource during the plan to verify that the object does not already exist.
```

## Allowing Empty Input for Attributes

> Migrated resources in the Terraform ACI provider support null, unknown, and known states. Terraform supports three states for any value: null (missing), unknown ("known after apply"), and known. Previous SDKs did not support null and unknown states, which prevented differentiating between an empty value and a non-provided value. Because of this limitation, updating a string value to an empty string was not possible before the Plugin Framework.

Reviewer 4: ```suggestion
The attributes of migrated resources in the Terraform ACI provider now support null, unknown, and known states which can be used to reset values to empty or to enforce that children objects are not present. Terraform supports three states for any value: null (missing), unknown ("known after apply"), and known. Previous SDKs did not support null and unknown states, which prevented differentiating between an empty value and a non-provided value. Because of this limitation, updating a string value to an empty string, updating a map to an empty map or a list of maps to an empty list was not possible before the Plugin Framework but is now supported in the migrated resources.
```

    ```bash
    Terraform will perform the following actions:                            

    # aci_epg_to_domain.terraform_epg_to_domain will be destroyed
    # (because aci_epg_to_domain.terraform_epg_to_domain is not in configuration)
    - resource "aci_epg_to_domain" "terraform_epg_to_domain" {
      - application_epg_dn    = "uni/tn-terraform_tenant/ap-terraform_ap/epg-terraform_epg" -> null
      - tdn                   = "uni/phys-phys" -> null
      ...
      - vmm_id                = jsonencode({})
        # (7 unchanged attributes hidden)
    }

    Plan: 0 to add, 0 to change, 1 to destroy.
    ```

    The [terraform state rm](https://developer.hashicorp.com/terraform/cli/commands/state/rm) command `terraform state rm aci_epg_to_domain.terraform_epg_to_domain` should be used to unmanage without destroying a resource. The output should reflect the succesful removal of the resource from state.

    ```bash
    Removed aci_epg_to_domain.terraform_epg_to_domain
    Successfully removed 1 resource instance(s).
    ```

>     The `terraform plan` output should reflect no changes, when the resource is unmanaged.

Reviewer: Wouldn't terraform plan try to add the child resource resource back at this point?
Shall we mention that the child resource can be removed from the config file, once the resource is no longer managed?

Reviewer: Resolving as discussed
